File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        combinations = []
        def dfs(index, stack, current_sum):
            if current_sum == target:
                combinations.append(stack.copy())
                return

            if current_sum > target or index >= len(nums):
                return

            for idx in range(index, len(nums)):
                stack.append(nums[idx])
                dfs(idx, stack, current_sum + nums[idx])
                stack.pop()

            # stack.append(nums[index])
            # dfs(index, stack, current_sum + nums[index])
            # stack.pop()
            # dfs(index + 1, stack, current_sum)
        
        dfs(0, [], 0)
        return combinations

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
     List<List<Integer>> result = new ArrayList<>();
     dfs(candidates, target, new ArrayList<>(), result, 0);
     return result;
    }

    void dfs(int[] candidates, int target, List<Integer> stack, List<List<Integer>> result, int index) {
        //base case
        if(target < 0){
            return;
        }else if(target == 0){
            result.add(new ArrayList<>(stack));
            return;
        }

        for(int idx = index; idx < candidates.length; idx++){
            stack.add(candidates[idx]);
            dfs(candidates, target - candidates[idx], stack, result, idx);
            stack.remove(stack.size() - 1);
        }
    }
}